File: InterviewQuestion/largestProduct.js
/* Write a function that accepts an array of integers and returns the largest
number we can get by multiplying two of them. For example, for the array:

```[-100, 2, 42, 100]
```

...We should get `42 * 100 = 4200`. */

// Naive O(n^2)

function largestMultiple(arr) {
  if(!(arr instanceof Array)) {
    throw new Error("largestMultiple() only accepts array arguments");
  }
  let maxProduct = Number.NEGATIVE_INFINITY;
  arr.forEach((num, i) => {
    arr.forEach((num2, x) => {
        if(i !== x){
          maxProduct = Math.max(maxProduct, num * num2);
        }
    });
  });

  return maxProduct;
}

// Smart O(n)

function largestMultipleSmarter(arr) {
  if(!(arr instanceof Array)) {
    throw new Error("largestMultipleSmarter() only accepts array arguments");
  }

  let largestPos;
  let lrgPosIndex;
  let secondPos;

  let largestNeg;
  let lrgNegIndx;
  let secondNeg;

  let maxProductPos;
  let maxProductNeg;

  arr.forEach((each, i) => {

    if (each >= 0) {

      if (!largestPos) {
        largestPos = each;
        lrgPosIndex = i;

      } else if (each > largestPos) {
          secondPos = largestPos;
          largestPos = each;
          lrgPosIndex = i;

      } else if ((!secondPos || each > secondPos) && i !== lrgPosIndex) {
          secondPos = each;
      }

    } else {

      if (!largestNeg) {
        largestNeg = each;
        lrgNegIndx = i;

      } else if (each < largestNeg) {
          secondNeg = largestNeg;
          largestNeg = each;
          lrgNegIndx = i;

      } else if ((!secondNeg || each < secondNeg) && i !== lrgNegIndx) {
          secondNeg = each;
      }

    }

  });

  if (largestPos && secondPos) {
    maxProductPos = largestPos * secondPos;
  }

  if (largestNeg && secondNeg) {
    maxProductNeg = largestNeg * secondNeg;
  }

  if (maxProductNeg && maxProductPos) {
    return Math.max(maxProductPos, maxProductNeg);
  } else if (maxProductPos) {
    return maxProductPos;
  } else {
    return maxProductNeg;
  }

}

module.exports = { largestMultiple, largestMultipleSmarter };

File: InterviewQuestion/largestProduct.py
'''
/* Write a function that accepts an array of integers and returns the largest
number we can get by multiplying two of them. For example, for the array:

[-100, 2, 42, 100]


...We should get `42 * 100 = 4200`. */

'''
import math

# Naive O(n^2)

def largest_multiple(arr):

    max_product = float('-inf')

    for i, val in enumerate(arr):
        for j, val2 in enumerate(arr):
            if i != j:
                max_product = max(max_product, val * val2)

    return max_product;

# Smart O(n)

def largest_multiple_smarter(arr):

    largestPos = None
    lrgPosIndex = None
    secondPos = None

    largestNeg = None
    lrgNegIndx = None
    secondNeg = None

    maxProductPos = None
    maxProductNeg = None

    for i, val in enumerate(arr):
        if val >= 0:
            if not largestPos:
                largestPos = val
                lrgPosIndex = i

            elif val > largestPos:
                secondPos = largestPos
                largestPos = val
                lrgPosIndex = i

            elif (not secondPos or val > secondPos) and i != lrgPosIndex:
                secondPos = each

        else:

            if not largestNeg:
                largestNeg = val
                lrgNegIndx = i

            elif val < largestNeg:
                secondNeg = largestNeg
                largestNeg = val
                lrgNegIndx = i

            elif ((not secondNeg or (val < secondNeg)) and i != lrgNegIndx):
                secondNeg = each


    if largestPos and secondPos:
        maxProductPos = largestPos * secondPos

    if largestNeg and secondNeg:
        maxProductNeg = largestNeg * secondNeg

    if maxProductNeg and maxProductPos:
        return max(maxProductPos, maxProductNeg)
    elif maxProductPos:
        return maxProductPos
    else:
        return maxProductNeg


print(largest_multiple_smarter([-100, 2, 42, 100]))
